### PR TITLE
Add automatic publishing of wheel and docker

### PR DESCRIPTION
**Pull request by ad-daniel**
This PR adds automatic build and publishing of python wheel and docker images.

https://github.com/opendr-eu/opendr/pull/200 needs to be merged first.

Just for testing at this moment.

### CODE REVIEW

**Comment by ad-daniel** [comment]
@passalis the wheel push worked fine. Can you double-check the tags of the docker ? I don't believe there's any way of testing for dockerhub, so I can only push once and delete them manually I think.

File: .github/workflows/publisher.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        twine upload --repository testpypi dist/*
+  publish-docker-cpu:
+    if: ${{ contains(github.event.pull_request.labels.*.name, 'tmp') }}
+    runs-on: ubuntu-20.04
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        ref: feature-automatic-publish # to remove before merge
+    - name: Build Docker Image
+      run: docker build --tag opendr-toolkit:cpu_latest --file Dockerfile .
```
Comment (passalis) [inline .github/workflows/publisher.yml:55]:
```suggestion
      run:  |
      OPENDR_VERSION=$(python3 -c "exec(open('src/opendr/_version.py').read()); print(__version__)")
      docker build --tag opendr-toolkit:cpu_v$OPENDR_VERSION --file Dockerfile .
```

File: .github/workflows/publisher.yml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    - name: Login to Docker Hub
+      uses: docker/login-action@v1
+      with:
+        username: ${{ secrets.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_PASSWORD }}
+    - name: Publish Image
+      run: |
+        docker tag opendr-toolkit:cpu_latest opendr/opendr-toolkit:cpu_latest
```
Comment (passalis) [inline .github/workflows/publisher.yml:63]:
```suggestion
        docker tag opendr-toolkit:cpu_v$OPENDR_VERSION opendr/opendr-toolkit:cpu_v$OPENDR_VERSION 
        docker tag opendr-toolkit:cpu_v$OPENDR_VERSION opendr/opendr-toolkit:cpu_latest
docker```

File: .github/workflows/publisher.yml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    - name: Build Docker Image
+      run: docker build --tag opendr-toolkit:cuda_latest --file Dockerfile-cuda .
+    - name: Login to Docker Hub
+      uses: docker/login-action@v1
+      with:
+        username: ${{ secrets.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_PASSWORD }}
+    - name: Publish Image
+      run: |
+        docker tag opendr-toolkit:cuda_latest opendr/opendr-toolkit:cuda_latest
```
Comment (passalis) [inline .github/workflows/publisher.yml:81]:
```suggestion
        docker tag opendr-toolkit:cuda_v$OPENDR_VERSION opendr/opendr-toolkit:cuda_v$OPENDR_VERSION 
        docker tag opendr-toolkit:cuda_v$OPENDR_VERSION opendr/opendr-toolkit:cuda_latest
```

File: .github/workflows/publisher.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        #docker push opendr/opendr-toolkit:cpu_latest
+  publish-docker-cuda:
+    if: ${{ contains(github.event.pull_request.labels.*.name, 'tmp') }}
+    runs-on: ubuntu-20.04
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        ref: feature-automatic-publish # to remove before merge
+    - name: Build Docker Image
+      run: docker build --tag opendr-toolkit:cuda_latest --file Dockerfile-cuda .
```
Comment (passalis) [inline .github/workflows/publisher.yml:73]:
```suggestion
      run: |
         OPENDR_VERSION=$(python3 -c "exec(open('src/opendr/_version.py').read()); print(__version__)")
          docker build --tag opendr-toolkit:cuda_v$OPENDR_VERSION --file Dockerfile-cuda .
```

**Comment by passalis** [comment]
Seems fine! I think the images are overwritten automatically. So, I suggested a version-based tagging in order to also provide the previous versions in an easily accessible way, as we discussed.

**Comment by ad-daniel** [comment]
Good idea, you're right. It might be easier to read the version directly from the github tag created, which supposedly should always be the same as the one written in `_version.py` and follow the format: `vX.Y`. If github doesn't complain about the syntax I'll try to make a push. It will require creating a new release on github though, just so you know

**Comment by ad-daniel** [comment]
PS: I meant I'll try when https://github.com/opendr-eu/opendr/pull/200 is merged

**Comment by ad-daniel** [comment]
@passalis for whatever reason the event/workflow doesn't appear to be triggering (when publishing a new release with tag v1.0.2). It's odd because I've tested the same trigger on a different repo and there it worked nicely. It might be because I'm trying to make a release from a branch other than master, or I don't know. Could you try to? Just to rule out it isn't something weird like a roles issue.

**Comment by ad-daniel** [comment]
nevermind, we have the same roles, that can't be it

**Comment by ad-daniel** [comment]
The push works as intended (both wheel and docker), it wasn't working because I forgot to remove the `tmp` label condition I used for testing.
The workflow starts when a new release is published on github. The only thing that needs some care is to correctly tag the release with something like: v1.0.2 since this is what is used to tag the dockers (the wheel is tagged from `_version.py`.

I've cleaned up the published dockers already, and removed references to this branch in the workflow. The PR is therefore ready to be reviewed.

**Review by omichel — DISMISSED** [review]

**Review by passalis — DISMISSED** [review]
Seems fine to me as well! Thank you!!
I have only one question/comment: Should this workflow upload to testpypi or to (main) pypi? It seems that the uploading command corresponds to the main pypi repo, but we pass the secrets for the testpypi. Is this intentional or do we actually upload to testpypi?

**Comment by ad-daniel** [comment]
ops, you're right, [test.pypi.org/](test.pypi.org/) was just for test, normally we should push to [pypi.org/](pypi.org/), I forgot to update the credentials as well.
After I prepare the workflow for the testing of wheels created from a branch (which we would run before creating the release) we can be fairly certain things are working as intended so we can likely push directly.

**Review by omichel — APPROVED** [review]

**Review by passalis — APPROVED** [review]
Thank you!

**Pull request merged**